README.md
<!--
This file is part of Inspirehep.
Copyright (C) 2016 CERN.

Inspirehep is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

Inspirehep is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with Inspirehep; if not, write to the
Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
MA 02111-1307, USA.

In applying this license, CERN does not
waive the privileges and immunities granted to it by virtue of its status
as an Intergovernmental Organization or submit itself to any jurisdiction.
-->

# json-merger

[![Coverage Status](https://img.shields.io/coveralls/inspirehep/json-merger.svg)](https://coveralls.io/r/inspirehep/json-merger)
[![GitHub tag](https://img.shields.io/github/tag/inspirehep/json-merger.svg)](https://github.com/inspirehep/json-merger/releases)
[![PyPI downloads](https://img.shields.io/pypi/dm/json-merger.svg)](https://pypi.python.org/pypi/json-merger)
[![License](https://img.shields.io/github/license/inspirehep/json-merger.svg)](https://github.com/inspirehep/json-merger/blob/master/LICENSE)

Module for merging JSON Objects.

* Free software: GPLv2 license
* Documentation: https://pythonhosted.org/json-merger/

## Installation

This project uses [Poetry](https://python-poetry.org/) for dependency
management. Assuming you have Poetry installed, this is how you set up
your fork for local development:

```console
$ cd json-merger/
$ poetry install --with tests --extras contrib
```

You can now run the tests with:

```console
$ poetry run pytest
```


CHANGES.rst
..
    This file is part of Inspirehep.
    Copyright (C) 2016, 2017, 2018 CERN.

    Inspirehep is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Inspirehep is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Inspirehep; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.


Changes
=======

Version 0.7.5
--------------
- In dict_merger head is always kept if root matches update.
- In dict_merger update is always kept if root matches head.

Version 0.7.4
--------------
- ``KEEP_UPDATE_CONFLICT_ON_HEAD_DELETE`` doesn't take the root into account any longer.

Version 0.7.3
--------------

- Add new typoe of conflict `INSERT`.
- `KEEP_HEAD_ENTITIES_CONFLICT_ON_NEW_UPDATE` will create INSERT conflict instead of REMOVE one.

Version 0.7.2
--------------

- Add new strategy `KEEP_HEAD_ENTITIES_CONFLICT_ON_NEW_UPDATE` which will create conflict when there is a new value in update.

Version 0.7.1
--------------

- Fix bug when DictMergeOps.keep_longest is used inside lists.

Version 0.7.0
--------------

- Add a new strategy for conflicts on head delete (contrib).

Version 0.6.1
--------------

- Smarter handling of conflicts in case of authors with the same name (contrib).

Version 0.6.0
--------------

- Correctly handle unicode in author names (contrib).

Version 0.5.2:
--------------

- Fix duplicate patches.

Version 0.5.1:
--------------

- Fix ``patch_to_conflict_set`` for list patches.

Version 0.5.0:
--------------

- Conflict method ``to_json`` it returns a list of patches conflicts of a single one.

Version 0.4.0:
--------------

- It's now possible to have field- and content- dependent merger operations.

Version 0.3.2:
--------------

- Initial public release.


json_merger/config.py
# -*- coding: utf-8 -*-
#
# This file is part of Inspirehep.
# Copyright (C) 2016 CERN.
#
# Inspirehep is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Inspirehep is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Inspirehep; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.


class DictMergerOps(object):
    """Possible strategies for merging two base values.

    Attributes:
        FALLBACK_KEEP_HEAD: In case of conflict keep the `head` value.

        FALLBACK_KEEP_UPDATE: In case of conflict keep the `update` value.
    """
    allowed_ops = [
        'FALLBACK_KEEP_HEAD',
        'FALLBACK_KEEP_UPDATE'
    ]

    @staticmethod
    def keep_longest(head, update, down_path):
        """Keep longest field among `head` and `update`.
        """
        if update is None:
            return 'f'
        if head is None:
            return 's'

        return 'f' if len(head) >= len(update) else 's'


for mode in DictMergerOps.allowed_ops:
    setattr(DictMergerOps, mode, mode)


class UnifierOps(object):
    """
    Attributes:
        KEEP_ONLY_HEAD_ENTITIES: Merge entities in `update` with their match
            in `head` having as a base the match in `root`.

        KEEP_ONLY_UPDATE_ENTITIES: Merge entities in 'head' with their match
            in `update` having as a base the match in `root`.

        KEEP_UPDATE_AND_HEAD_ENTITIES_HEAD_FIRST: Perform an union of all
            entities from `head` and `update` and merge the matching ones.
            Also, preserve the order relations between the entities in both
            lists. If two entities can have the same position first pick the
            one that is present in the `head` object.

        KEEP_UPDATE_AND_HEAD_ENTITIES_UPDATE_FIRST: Same behavior as
            KEEP_UPDATE_AND_HEAD_ENTITIES_HEAD_FIRST but first pick the
            `update` entities.

        KEEP_UPDATE_ENTITIES_CONFLICT_ON_HEAD_DELETE: If an entity was added
            in the diff between the `root` and `head` lists but it's not
            present in the `update` list then raise a conflict.
    """
    allowed_ops = [
        'KEEP_ONLY_HEAD_ENTITIES',
        'KEEP_ONLY_UPDATE_ENTITIES',
        'KEEP_UPDATE_AND_HEAD_ENTITIES_HEAD_FIRST',
        'KEEP_UPDATE_AND_HEAD_ENTITIES_UPDATE_FIRST',
        'KEEP_UPDATE_ENTITIES_CONFLICT_ON_HEAD_DELETE',
        'KEEP_UPDATE_AND_HEAD_ENTITIES_CONFLICT_ON_HEAD_DELETE',
        'KEEP_HEAD_ENTITIES_CONFLICT_ON_NEW_UPDATE'
    ]


for mode in UnifierOps.allowed_ops:
    setattr(UnifierOps, mode, mode)
